### PR TITLE
feature: add queue_length gauge metrics for exporter

### PR DESCRIPTION
**Pull request by homholueng**
Hi Dani, as promised I am making a PR to add queue_length metric for exporter.

The implementation is simple. A daemon thread will start with main thread and it will scrape length of queues which declare by developer.

Looking forward to your opinions. 

Unittest will be added after code review.

Thx.

### CODE REVIEW

**Comment by homholueng** [comment]
> Looking good. Are there any other metrics we can get from .queue_declare() ?

According to [kombu API Refrence](https://docs.celeryq.dev/projects/kombu/en/master/reference/kombu.transport.qpid.html?highlight=queue_declare#kombu.transport.qpid.Transport.Connection.Channel.queue_declare), `queue_declare` returns

> A named tuple representing the declared queue as a named tuple. The tuple values are ordered as `queue`, `message count`, and the `active consumer count`.

We can add `Gause(active_consumer_count){label=queuename}` metric to exporter if it's necessary.

**Comment by danihodovic** [comment]
Could you add the active consumer count? That would be awesome :)

**Comment by homholueng** [comment]
`celery_active_consumer_count` is added.

**Comment by danihodovic** [comment]
Would it be possible to get the queue length and consumer count for all queues, without having to specify them as options in the exporter cli?

**Comment by homholueng** [comment]
> Would it be possible to get the queue length and consumer count for all queues, without having to specify them as options in the exporter cli?

The reasons for doing this are as follows: 

1. The lower kombu API have to **declare queue first** and then fetch the information of queue. So the specify of queues from developer is required. 
2. Maybe there is a way to auto discover queues in broker. But I think not every queue in broker is used by celery or by celery of "current developer". There is possible that the broker is used by many developer's celery instance so the data of every queues in the broker may cause confused. e.g. Celery will create queues like `celeryev.xxxx` to transport worker's event message, and these queues should not be concerned by developer.

So, I think the specify from developer is more reasonable and Explicit.

> Could you have a second look that it works? I believe the test should also show a consumer count of 1.

Sure. I have checked the [source code of kombu](https://github.com/celery/kombu/blob/master/kombu/transport/qpid.py#L680) and the [documentation of rabbitmq](https://www.rabbitmq.com/api-guide.html#passive-declaration), this metric does reflect the number of consumers. And I also check it on my host. It's really works. It's there any possible that the consumer process is already down when you read the metrics?
image

> Perhaps we could scrape queue length before building the output:

Yep. But after that, the pull request will become a non-constant complexity operation. It will affect by the length of queues which is `O(n)` complexity. It's acceptable?

**Comment by danihodovic** [comment]
> The lower kombu API have to declare queue first and then fetch the information of queue. So the specify of queues from developer is required.

I managed to run a breakpoint() in the test suite to inspect the celery app. The app has an inspect utility for listing queues. This seems to work for Redis, but you can try it for RabbitMQ to see if it shows the same data. I set the breakpoint() here https://github.com/danihodovic/celery-exporter/blob/master/src/exporter.py#L201

```
self.app.control.inspect().active_queues()

{
	'gen89100@x1': [
		{
			'name': 'celery',
			'exchange': {'name': 'celery', 'type': 'direct', 'arguments': None, 'durable': True, 'passive': False, 'auto_delete': False, 'delivery_mode': None, 'no_declare': False},
			'routing_key': 'celery',
			'queue_arguments': None,
			'binding_arguments': None,
			'consumer_arguments': None,
			'durable': True,
			'exclusive': False,
			'auto_delete': False,
			'no_ack': False,
			'alias': None,
			'bindings': [],
			'no_declare': None,
			'expires': None,
			'message_ttl': None,
			'max_length': None,
			'max_length_bytes': None,
			'max_priority': None
		}
	]
}

```
This should give us a list of queues for each worker. We can flatten the list and then get the length for each.

> Sure. I have checked the [source code of kombu](https://github.com/celery/kombu/blob/master/kombu/transport/qpid.py#L680) and the [documentation of rabbitmq](https://www.rabbitmq.com/api-guide.html#passive-declaration), this metric does reflect the number of consumers.

In that case shouldn't the unit test you added show 1 consumer instead of 0? https://github.com/danihodovic/celery-exporter/pull/118/files#diff-fb60222581054c3ac15e1981679b0a596c22dfeb483873337979183826f559ffR76. In any case this is not a deal breaker.

> Yep. But after that, the pull request will become a non-constant complexity operation. It will affect by the length of queues which is O(n) complexity. It's acceptable?

That's fine. Given that there is a finite set of queues (usually less than 10) I don't think reading queued data for all queues will have an outsized performance impact. If it does, we can cache the values:

> If a metric is particularly expensive to retrieve, i.e. takes more than a minute, it is acceptable to cache it. This should be noted in the HELP string.

https://prometheus.io/docs/instrumenting/writing_exporters/#scheduling

**Comment by homholueng** [comment]
Thanks for your advice.

According to the previous discussion, I will make the following modification:

1. Remove `--track-queue` and `--queue-track-interval` option and fetch track queue list by `celery_app.inspect().active_queues`.
2. Remove daemon thread. `celery_queue_length` and `celery_active_consumer_count ` metrics scrape only perform at [metrics pulling](https://github.com/danihodovic/celery-exporter/blob/1534d70086becec29b9f0833115a1f05a620f4e9/src/http_server.py#L33)
3. Add non-zero `celery_active_consumer_count` metrics unit test case.

Those modifications will be committed ASAP.

**Comment by homholueng** [comment]
Something unexpected happened while I was adding non-zero `celery_active_consumer_count` metrics unit test case. 

It seems the `memory://localhost` broker type does not support follow the amqp protocol so the `queue_declare` method can not get the correct `celery_active_consumer_count`.

So I also verified it with redis as broker and can not get the correct `celery_active_consumer_count` too.

Should we keep the `celery_active_consumer_count` metric?

**Comment by homholueng** [comment]
> Something unexpected happened while I was adding non-zero `celery_active_consumer_count` metrics unit test case.
> 
> It seems the `memory://localhost` broker type does not follow the amqp protocol so the `queue_declare` method can not get the correct `celery_active_consumer_count`.
> 
> So I also verified it with Redis as broker and can not get the correct `celery_active_consumer_count` too.
> 
> Should we keep the `celery_active_consumer_count` metric?

I dig into the `kombu` implementation and find some clue.

`queue_declare` will raise `amqp.exceptions.ChannelError: Channel.queue_declare: (404) NOT_FOUND - no queue 'celery' in vhost '/'` when there is no message in Redis broker because the "queue" in Redis is a `list` and it seems not durable. (The key will be deleted when the queue is empty)

**Comment by danihodovic** [comment]
If the queue is empty there is no active consumer, have I understood this right? In that case we could set it to zero.

**Comment by homholueng** [comment]
> If the queue is empty there is no active consumer, have I understood this right? In that case we could set it to zero.

Yep, but I also test with non-empty Redis queue and alive worker. The `queue_declare` method didn't raise but `celery_active_consumer_count` is still 0. 

So, I guess only broker that support amqp protocol will return consumer count.

**Comment by danihodovic** [comment]
What is your preference? Keep if for RabbitMQ or remove it so that it's uniform for all brokers?

**Comment by homholueng** [comment]
> What is your preference? Keep if for RabbitMQ or remove it so that it's uniform for all brokers?

I think we can keep the metric as long as we explain the limit of broker type in the doc.

**Comment by danihodovic** [comment]
Sounds good, maybe we can make a PR to Kombu to add the functionality for Redis.

**Comment by homholueng** [comment]
> Is the PR ready for review?

Yes

File: src/test_cli.py [hunk 1]
```diff
@@ -34,49 +35,56 @@ def succeed():
...
-    )
-    assert (
-        f'celery_task_runtime_count{{hostname="{celery_worker.hostname}",name="src.test_cli.succeed"}} 2.0'
-        in res.text
-    )
+    # start worker and consume message in broker
+    with start_worker(celery_app, without_heartbeat=False) as celery_worker:
+        time.sleep(2)
+        res = requests.get("http://localhost:23000/metrics")
+        assert res.status_code == 200
```
Comment (danihodovic) [inline src/test_cli.py:59]: Can we un-indent the asserts?

**Review by danihodovic — CHANGES_REQUESTED** [review]
Minor nit, rest looks good :)

**Comment by danihodovic** [comment]
Nice work @homholueng :clap: 伟大的工程工作

**Comment by danihodovic** [comment]
Released as danihodovic/celery-exporter:0.5.0

Can you update README.md to add the new metrics? https://github.com/danihodovic/celery-exporter#metrics

**Comment by homholueng** [comment]
> https://github.com/danihodovic/celery-exporter#metrics

Sure
